docs/remote-attestation.md

```diff
@@ -1,212 +1,111 @@
+participant T as Trusted Enclave
+participant S as Secure Processor
+
+Note over U,S: Boot Trusted Enclave
+
+U->>S: Trusted Enclave Binary + firmware
+S->>T: Boot Trusted Enclave
+
+Note over U,S: Initialize
+
+U->>T: Configuration
+activate T
+T-->>T: Init logic
+T-->>T: Generate Encryption key pair
+T-->>T: Put Encryption public key hash in AttestationReport request
+T->>S: AttestationReportRequest
+S-->>S: Generate AttestationReport
+S->>T: Signed AttestationReport
+T->>U: enclave Public Key
+deactivate T
+
+Note over C,S: Fetch Enclave Public Key
+
+C->>U: GetPublicKeyRequest
+activate U
+U->>C: GetPublicKeyResponse
+deactivate U
+C-->>C: Verify Enclave Public Key attestation
+
+Note over C,S: Send Encrypted Data
+
+loop
+   C-->>C: encrypt request
+   C->>U: InvokeRequest
+   activate U
+   U->>T: InvokeRequest
+   activate T
+   T-->>T: decrypt request
+   T-->>T: process request
+   T-->>T: encrypt response
+   T->>U: InvokeResponse
+   deactivate T
+   U->>C: InvokeResponse
+   deactivate U
+   C-->>C: decrypt response
+end
+```
 
-  Trusted Runtime-->>Trusted Runtime: Generate a Signing key pair
-  Trusted Runtime-->>Trusted Runtime: Put Signing public key hash in the AttestationReport request
-  Trusted Runtime->>TEE Platform: AttestationReport request
-  TEE Platform-->>TEE Platform: Generate AttestationReport
-  TEE Platform->>Trusted Runtime: Signed AttestationReport
+### Fetch Enclave Public Key
 
-  Note over Untrusted Launcher,Trusted Runtime: Initialization
+Initially, the client connects to the untrusted launcher and requests the
+attested enclave public key. This key was generated in the enclave, but it is
+then kept in memory by the corresponding untrusted launcher, so it may be served
+to the client without having to reach into the enclave every time.
 
-  Untrusted Launcher->>Trusted Runtime: Configuration and business logic
+The client then checks:
 
-  Client-->>Client: Generate a Signing key pair
+- that the attestation report is signed by the TEE manufacturer (e.g. AMD,
+  Intel)
+- that the attestation report additional data includes the hash of the public
+  key, to confirm that the key pair was in fact generated from inside the
+  enclave
+- that the attestation report measurement corresponds to a trusted version of
+  the enclave binary
 
-  Note over Client,Trusted Runtime: Handshake Start
+If any of these checks fails, the client refuses to go ahead.
 
-  Client-->>Client: Generate an Ephemeral key pair
-  Client->>Untrusted Launcher: ClientHello
-  Untrusted Launcher->>Trusted Runtime: ClientHello
-  Trusted Runtime-->>Trusted Runtime: Generate an Ephemeral key pair
+### Send Encrypted Data
 
-  Trusted Runtime->>Untrusted Launcher: ServerIdentity
-  Untrusted Launcher->>Client: ServerIdentity
-  Client-->>Client: Validate ServerIdentity
+The client also generates an encryption key pair ahead of time (it may be reused
+across invocations, or it may be generated for each invocation).
```

Comment:
> reused across invocations

This could lead to subtle issues.
Say Client A uses the same pk_A in requests to enclaves T1 and T2, and each respond with messages m1 and m2 encrypted with pk_A
The attacker can make it appear that T1 responded with m2 and T2 with m1. 

This may not break certain kinds of applications, but it makes it easier to reason about security for **all** applications following this paradigm if clients require per-request response keys.

Reply:
Good point, changed to use fresh symmetric keys for each invocation, PTAL.